opte/src/engine/rule.rs

```diff
@@ -1216,6 +1244,36 @@ pub struct RulePredicates {
     data_preds: Vec<DataPredicate>,
 }
 
+impl PartialEq for RulePredicates {
+    /// Rule predicates are equal when both contain identical sets of
+    /// header and data predicates.
+    fn eq(&self, other: &Self) -> bool {
+        if self.hdr_preds.len() != other.hdr_preds.len() {
+            return false;
+        }
+
+        if self.data_preds.len() != other.data_preds.len() {
+            return false;
+        }
+
+        for hp in &self.hdr_preds {
+            if !other.hdr_preds.contains(hp) {
```

Comment:
This is probably a _tiny_ issue now, but we may want to store the predicates in a way that avoid this O(N**2) comparison here. If we can sort the predicates, you could use [`binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) to get O(N log N) or just compare hashes to get O(2 * N) == O(N).

Reply:
Yea, honestly I'm not too worried about this atm with everything else. It also makes me thing the first thing we might want is to have an upper bound on number of predicates as that should probably be a thing no matter what algo we use.

Reply:
Yeah, I figured it'd be lost in the noise of other stuff. Thanks!